### PR TITLE
Fix the rect elemets position when drawing svgs

### PR DESCRIPTION
**Pull request by MatheusrdSantos**

### CODE REVIEW

**Comment by rocinante-sys** [comment]
@MatheusrdSantos I patched my pdf-lib with your code and it is better, but still does not produce correct SVG in React. I see that the SVG is streatched in Y axis.

**Comment by MatheusrdSantos** [comment]
Hi, @rocinante-sys.
Sorry for my late response. Could you provide an example of the SVG you are using and the result you got?

File: src/api/svg.ts [hunk 1]
```diff
@@ -76,6 +79,19 @@ interface SVGElementToDrawMap {
   [cmd: string]: (a: SVGElement) => Promise<void>;
 }
 
+const computeDotDistance = (d1: Coordinates, d2: Coordinates) => {
```
Comment (Sharcoux) [inline src/api/svg.ts:82]: Use distanceCoords

File: src/api/svg.ts [hunk 2]
```diff
@@ -76,6 +79,19 @@ interface SVGElementToDrawMap {
...
+  return Math.sqrt((d1.x - d2.x) ** 2 + (d1.y - d2.y) ** 2)
+}
+
+const inRange = (v: number, min: number, max: number) => {
```
Comment (Sharcoux) [inline src/api/svg.ts:86]: What does it means, and what are those parameters ? What is this goal of this function ? This should normally be one of:

```js
const isInRange = (x, min, max) => min <=x && x<=max
const inRange = (x, min, max) => Math.max(Math.min(x, max), min)
```

File: src/api/svg.ts [hunk 3]
```diff
@@ -76,6 +79,19 @@ interface SVGElementToDrawMap {
...
+  return Math.min(Math.trunc(v), Math.trunc(min)) === Math.trunc(min) && Math.max(Math.trunc(v), Math.trunc(max)) === Math.trunc(max)
+}
+
+const isCoordinateInsideTheRect  = (dot: Coordinates, rect: RectangleTranslatable, padding = 1) => {
```
Comment (Sharcoux) [inline src/api/svg.ts:90]: Use `isEqual(0, distance(dot, rect.orthoProjection(dot)))` instead

File: src/api/svg.ts [hunk 4]
```diff
@@ -112,29 +129,60 @@ const runnersToPage = (
...
-          ? options.fonts[element.svgAttributes.fontFamily]
-          : undefined,
-      size: fontSize,
-      color: element.svgAttributes.fill,
-      opacity: element.svgAttributes.fillOpacity,
-      rotate: element.svgAttributes.rotate,
-    });
+    const point = { x: (element.svgAttributes.x || 0) - offset, y:  element.svgAttributes.y || 0 }
+
+    if (isCoordinateInsideTheRect(point, svgRect)) {
```
Comment (Sharcoux) [inline src/api/svg.ts:134]: From what I understand, when the text is not inside the bounds, it won't show up, right? Then, let pdf-lib handle this and for now don't worry about it. If you really want to do sth here, that would be to cut the text and draw only the letters that are inside the bound.

File: src/utils/converters.ts [hunk 5]
```diff
@@ -0,0 +1,49 @@
+import { Converter, Coordinates, Size, Space } from '../types'
+
+import { norm } from './maths'
```
Comment (Sharcoux) [inline src/utils/converters.ts:3]: Don't bring the WHOLE code of graph here. Just the minimal part that you need. This is probably the functions, and the basic classes. You just need an object able to store the details about an element, and make basic calculations about it (like orthoProjection)

**Review by Sharcoux — CHANGES_REQUESTED** [review]

File: src/api/svg.ts (hunk 2 above)
Comment (MatheusrdSantos) [reply src/api/svg.ts:86]: indeed, I overcomplicated this function

File: src/api/svg.ts (hunk 4 above)
Comment (MatheusrdSantos) [reply src/api/svg.ts:134]: That's true for the pdf's boundings, but not for the viewBox's bounding.
I agree that this implementation is incomplete, but it's enough for now.

File: src/api/svg.ts [hunk 6]
```diff
@@ -112,38 +117,169 @@ const runnersToPage = (
...
+      // if there's no intersection it means that the line doesn't intersects the svgRect and isn't visible
+      if (intersection.length === 0) return
+
+      if (!isStartInside) {
+        // replace the line start point by the nearest intersection
+        const nearestPoint = intersection.sort((p1, p2) => distanceCoords(start, p1) - distanceCoords(start, p2))[0]
+        start = new PointXY(nearestPoint)
+      }
+
+      if (!isEndInside) {
```
Comment (Sharcoux) [inline src/api/svg.ts:166]: You are repeating yourself here

File: src/api/svg.ts [hunk 7]
```diff
@@ -112,38 +117,169 @@ const runnersToPage = (
...
+            })
+            const normalizedNext = normalizePoint(nextPoint)
+
+            let endPoint = new PointXY({ x: nextPoint.x, y: nextPoint.y })
+            let startPoint = new PointXY({ x: currentPoint.x, y: currentPoint.y })
+
+            const isStartInside = isCoordinateInsideTheRect(startPoint, svgRect)
+            const isEndInside = isCoordinateInsideTheRect(endPoint, svgRect)
+
+            if (!(isStartInside && isEndInside)) {
```
Comment (Sharcoux) [inline src/api/svg.ts:224]: Feels like you are repeating the code from the line. Can't you create a function that takes a line and the viewbox and returns the inner part of the line?

File: src/api/svg.ts [hunk 8]
```diff
@@ -596,6 +732,11 @@ const parseAttributes = (
     const inner = newInherited.fontFamily.match(/^"(.*?)"|^'(.*?)'/);
     if (inner) newInherited.fontFamily = inner[1] || inner[2];
   }
+  
+  if (newInherited.strokeWidth) {
+    newInherited.strokeWidth = newConverter.size(newInherited.strokeWidth, newInherited.strokeWidth).width
```
Comment (Sharcoux) [inline src/api/svg.ts:737]: Maybe you should use the max between (xx.width, xx.height, 1), xx being the result of newConverter.size()

File: src/types/index.ts [hunk 9]
```diff
@@ -1 +1,77 @@
...
+
+export interface Resizable {
+  getHandles(): (Point & Translatable)[]
+}
+
+export interface Linkable {
+  getLinks(): Point[]
+}
+
+export interface Editable {
```
Comment (Sharcoux) [inline src/types/index.ts:71]: Useless

File: src/types/index.ts [hunk 10]
```diff
@@ -1 +1,77 @@
...
+  space: Space
+  measurePrecision: number
+  anglePrecision: number
+}
+
+export interface Translatable {
+  translate(vector: Coordinates): void
+}
+
+export interface Rotatable {
```
Comment (Sharcoux) [inline src/types/index.ts:58]: Useless

File: src/types/index.ts [hunk 11]
```diff
@@ -1 +1,77 @@
...
+  fontSize: number
+  strokeScale: number
+  toPixelConverter: Converter
+  toRealConverter: Converter
+  space: Space
+  measurePrecision: number
+  anglePrecision: number
+}
+
+export interface Translatable {
```
Comment (Sharcoux) [inline src/types/index.ts:54]: Useless

File: src/types/index.ts [hunk 12]
```diff
@@ -1 +1,77 @@
...
+  y: (y: number) => number
+  w: (w: number) => number
+  h: (h: number) => number
+  point: (coords: Coordinates) => Coordinates
+  norm: (vector: Coordinates) => number
+}
+
+export type LinkElement = Rectangle | Ellipse
+
+export type DrawConfig = {
```
Comment (Sharcoux) [inline src/types/index.ts:44]: Mostly useless

File: src/types/index.ts [hunk 13]
```diff
@@ -1 +1,77 @@
...
+  | Ellipse
+  | Line
+  | Plot
+  | Point
+  | Rectangle
+  | Segment
+
+  export type Space = { xMin: number; yMin: number; xMax: number; yMax: number }
+
+export type Converter = {
```
Comment (Sharcoux) [inline src/types/index.ts:33]: We already defined our converters here, using point, size, x, y, width and height

File: src/types/index.ts [hunk 14]
```diff
@@ -1 +1,77 @@
...
+
+export interface Linkable {
+  getLinks(): Point[]
+}
+
+export interface Editable {
+  setText(text: string): void
+}
+
+export interface Functional {
```
Comment (Sharcoux) [inline src/types/index.ts:75]: Useless

File: src/utils/elements/Arc.ts [hunk 15]
```diff
@@ -0,0 +1,147 @@
+import { angleABC, distance, distanceCoords, rotate, vector } from '../maths'
+
+import { CircleORay } from './Circle'
+import GraphElement from './GraphElement'
+import Point, { PointXY, Projected } from './Point'
+
+export default abstract class Arc extends GraphElement {
```
Comment (Sharcoux) [inline src/utils/elements/Arc.ts:7]: Are we using this somewhere?

File: src/utils/elements/Arc.ts [hunk 16]
```diff
@@ -0,0 +1,147 @@
...
+      const coords =
+        distanceCoords(H.toCoords(), origin) < distanceCoords(H.toCoords(), destination)
+          ? origin
+          : destination
+      return new PointXY(coords)
+    }
+  }
+}
+
+export class ArcOAB extends Arc {
```
Comment (Sharcoux) [inline src/utils/elements/Arc.ts:74]: Remove

File: src/utils/elements/Arc.ts [hunk 17]
```diff
@@ -0,0 +1,147 @@
...
+      this.center(),
+      this.destination(),
+      this.lastSweep
+    )
+    return this.lastSweep
+  }
+}
+
+/** Create an Arc that will be removed if the original support is removed, or the origin of the arc. */
+export class ArcDependant extends Arc {
```
Comment (Sharcoux) [inline src/utils/elements/Arc.ts:120]: Useless

File: src/utils/elements/Ellipse.ts [hunk 18]
```diff
@@ -0,0 +1,183 @@
...
+  times,
+  unitVector,
+  vector,
+} from '../maths'
+
+import GraphElement from './GraphElement'
+import Point, { PointXY } from './Point'
+import Segment, { SegmentAB } from './Segment'
+
+export default abstract class Ellipse extends GraphElement {
```
Comment (Sharcoux) [inline src/utils/elements/Ellipse.ts:18]: We just need one implementation of the ellipse, based on the data model we are using. No need for an abstract class + many implementations

File: src/utils/elements/Arc.ts (hunk 15 above)
Comment (MatheusrdSantos) [reply src/utils/elements/Arc.ts:7]: Currently, we aren't using it, but I think we might need to handle arc commands from svg paths in the future

**Pull request merged**